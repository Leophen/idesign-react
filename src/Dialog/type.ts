export interface DialogProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 控制对话框显示隐藏
   * @default false
   */
  visible?: boolean;
  /**
   * 按下退出键是否触发关闭事件
   * @default true
   */
  closeOnEsc?: boolean;
  /**
   * 是否显示遮罩层
   * @default true
   */
  showMask?: boolean;
  /**
   * 对话框宽度
   */
  width?: React.CSSProperties["width"];
  /**
   * 对话框头部内容
   */
  header?: React.ReactNode;
  /**
   * 对话框底部内容
   */
  footer?: React.ReactNode;
  /**
   * 对话框关闭时触发事件
   */
  onClose?: () => void;
}
