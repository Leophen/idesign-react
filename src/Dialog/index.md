---
nav:
  title: 组件
  path: /components
group:
  title: 弹框提示组件
  order: 5
order: 4
---

# Dialog 对话框

模态对话框。

## 基本用法

可通过 `visible` 属性控制对话框显示隐藏。

```tsx
import React, { useState } from 'react';
import { Button, Dialog } from 'idesign-react';

const App = () => {
  const [visible, setVisible] = useState(false);
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button type="primary" onClick={() => setVisible(true)}>
          打开对话框
        </Button>
        <Dialog header="对话框标题" visible={visible} onClose={() => setVisible(false)}>
          <span>对话框内容</span>
        </Dialog>
      </div>
    </div>
  );
};

export default App;
```

## 禁用退出键关闭功能

可通过 `closeOnEsc` 属性控制打开对话框时按下 ESC 退出键是否触发关闭事件，默认为 `true`，表示 ESC 退出键触发 `onClose` 关闭事件。

```tsx
import React, { useState } from 'react';
import { Button, Dialog } from 'idesign-react';

const App = () => {
  const [visible, setVisible] = useState(false);
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button type="primary" onClick={() => setVisible(true)}>
          禁用对话框退出键功能
        </Button>
        <Dialog
          header="对话框标题"
          visible={visible}
          closeOnEsc={false}
          onClose={() => setVisible(false)}
        >
          <span>对话框内容</span>
        </Dialog>
      </div>
    </div>
  );
};

export default App;
```

## 隐藏遮罩层

可通过 `showMask` 属性控制打开对话框时是否显示遮罩层，默认为 `true`。

```tsx
import React, { useState } from 'react';
import { Button, Dialog } from 'idesign-react';

const App = () => {
  const [visible, setVisible] = useState(false);
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button type="primary" onClick={() => setVisible(true)}>
          隐藏遮罩层
        </Button>
        <Dialog
          header="对话框标题"
          visible={visible}
          showMask={false}
          onClose={() => setVisible(false)}
        >
          <span>对话框内容</span>
        </Dialog>
      </div>
    </div>
  );
};

export default App;
```

## 自定义对话框宽度

可通过 `width` 属性控制对话框宽度。

```tsx
import React, { useState } from 'react';
import { Button, Dialog } from 'idesign-react';

const App = () => {
  const [visible, setVisible] = useState(false);
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button type="primary" onClick={() => setVisible(true)}>
          自定义宽度
        </Button>
        <Dialog header="对话框标题" visible={visible} width={800} onClose={() => setVisible(false)}>
          <span>对话框内容</span>
        </Dialog>
      </div>
    </div>
  );
};

export default App;
```

## 自定义对话框头部

可通过 `header` 属性控制对话框头部内容。

```tsx
import React, { useState } from 'react';
import { Button, Dialog, Icon } from 'idesign-react';

const header = (
  <>
    <Icon name="TipWarningFill" />
    <span>这是一个自定义标题</span>
  </>
);

const App = () => {
  const [visible, setVisible] = useState(false);
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button type="primary" onClick={() => setVisible(true)}>
          自定义头部
        </Button>
        <Dialog header={header} visible={visible} onClose={() => setVisible(false)}>
          <span>对话框内容</span>
        </Dialog>
      </div>
    </div>
  );
};

export default App;
```

## 自定义对话框底部

可通过 `footer` 属性控制对话框底部内容。

```tsx
import React, { useState } from 'react';
import { Button, Dialog, Icon } from 'idesign-react';

const App = () => {
  const [visible, setVisible] = useState(false);

  const footer = (
    <>
      <Button variant="outline" onClick={() => setVisible(false)}>
        我再想想
      </Button>
      <Button>去意已决</Button>
    </>
  );

  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button type="primary" onClick={() => setVisible(true)}>
          自定义底部
        </Button>
        <Dialog
          header="对话框标题"
          footer={footer}
          visible={visible}
          onClose={() => setVisible(false)}
        >
          <span>对话框内容</span>
        </Dialog>
      </div>
    </div>
  );
};

export default App;
```

<API />
