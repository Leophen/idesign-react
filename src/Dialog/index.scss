.i-dialog-wrapper {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 2500;
  width: 100vw;
  height: 100vh;
  pointer-events: none;
}

.i-dialog__mask {
  position: absolute;
  top: 0;
  left: 0;
  z-index: 2500;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.6);
  pointer-events: auto;
}

@keyframes iDialogOpen {
  0% {
    transform: scale(0.2) translate(-50%, -50%);
  }
  100% {
    transform: scale(1) translate(-50%, -50%);
  }
}

.i-dialog {
  position: absolute;
  top: 50%;
  left: 50%;
  z-index: 2501;
  box-sizing: border-box;
  width: 520px;
  max-height: 100%;
  overflow: auto;
  color: var(--i-font-2);
  font-size: 14px;
  background: var(--i-bg);
  border: 1px solid var(--i-border);
  border-radius: 6px;
  box-shadow: var(--i-shadow);
  animation: iDialogOpen 0.2s ease-out forwards;
  pointer-events: auto;
}

.i-dialog__close {
  position: absolute;
  top: 14px;
  right: 14px;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 50%;
  cursor: pointer;
  transition: all 0.1s ease-out;
  .i-icon {
    margin: 0;
  }
  &:hover {
    background: var(--i-bg-hover);
  }
}

.i-dialog__header,
.i-dialog__body,
.i-dialog__footer {
  box-sizing: border-box;
  padding: 16px;
}

.i-dialog__body {
  overflow: overlay;
  overflow: auto;
}

.i-dialog__header {
  display: flex;
  align-items: center;
  font-weight: 700;
  font-size: 16px;
  word-break: break-word;
  .i-icon {
    margin-right: 8px;
    font-size: 20px;
  }
  .icon-TipInfo {
    color: var(--i-primary);
  }
  .icon-TipCheckFill {
    color: var(--i-success);
  }
  .icon-TipWarningFill {
    color: var(--i-warning);
  }
  .icon-TipCloseFill {
    color: var(--i-error);
  }
}

.i-dialog__footer {
  width: 100%;
  text-align: right;
  .i-button + .i-button {
    margin-left: 10px;
  }
}
