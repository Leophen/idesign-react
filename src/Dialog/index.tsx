import React, { useEffect, useState, useRef } from 'react';
import classNames from 'classnames';
import './index.scss';
import Icon from '../Icon';
import Button from '../Button';
import Transition from '../Transition';
import ReactDOM from 'react-dom';
import { DialogProps } from './type';

// 获取触发对话框打开的 DOM 节点原位置
let clickOpenTarget: EventTarget | null;
let mousePosition: { x: number; y: number } | null;
const getClickPosition = (e: MouseEvent) => {
  clickOpenTarget = e.target
  mousePosition = {
    x: e.clientX,
    y: e.clientY,
  };
  setTimeout(() => {
    mousePosition = null;
  }, 100);
};
if (typeof window !== 'undefined' && window.document && window.document.documentElement) {
  document.documentElement.addEventListener('click', getClickPosition, true);
}

// 创建对话框容器
let dialogWrapper = document.querySelector('#i-dialog-wrapper')
if (!dialogWrapper) {
  dialogWrapper = document.createElement('div')
  dialogWrapper.className = 'i-dialog-wrapper'
  dialogWrapper.id = 'i-dialog-wrapper'
  document.body.append(dialogWrapper)
}

const Dialog: React.FC<DialogProps> = (props) => {
  const {
    children = '',
    className,
    style,
    visible = false,
    closeOnEsc = true,
    showMask = true,
    width,
    header,
    footer,
    onClose = () => { },
    ...restProps
  } = props;

  const handleKeyDown = (e: any) => {
    if (e.key === 'Escape') {
      onClose?.()
    }
  };

  const closeDialog = () => {
    onClose?.()
    closeOnEsc && document.removeEventListener('keydown', handleKeyDown)
  }

  // 判断点击节点是否父元素内
  const hasParent = (node: any, parent: HTMLElement) => {
    while (node) {
      if (node === parent) {
        return true;
      }
      node = node.parentNode
    }
    return false;
  };

  const handleClick = (e: any) => {
    if (!hasParent(e.target, dialog.current) && e.target !== clickOpenTarget) {
      closeDialog()
      document.removeEventListener('click', handleClick, true)
    }
  }

  const dialog = useRef<any>(null);

  // 打开对话框时禁止背景滚动，对原 overflow 进行备份
  const bodyOverflow = useRef<string>(document.body.style.overflow);

  useEffect(() => {
    if (visible) {
      // 打开对话框时禁止背景滚动
      document.body.style.overflow = 'hidden';
      // 退出键功能
      closeOnEsc && document.addEventListener('keydown', handleKeyDown)
      // 展开动画出发点
      if (mousePosition && dialog.current) {
        dialog.current.style.transformOrigin = `${mousePosition.x - dialog.current.offsetLeft}px ${mousePosition.y - dialog.current.offsetTop
          }px`;
      }
      // 无遮罩层时点击关闭功能
      !showMask && document.addEventListener('click', handleClick, true)
    } else {
      // 关闭对话框时恢复背景滚动
      document.body.style.overflow = bodyOverflow.current;
    }
  }, [visible])

  const DialogNode = (
    <>
      {showMask && <Transition
        timeout={200}
        in={visible}
        animation='fade-in'
        key='i-dialog__mask'
      >
        <div className="i-dialog__mask" onClick={closeDialog} onScroll={() => { return }}></div>
      </Transition>}

      <Transition
        timeout={200}
        in={visible}
        animation='fade-in'
        key='i-dialog'
      >
        <div
          ref={dialog}
          className={classNames(
            'i-dialog',
            className
          )}
          style={{ ...(style || {}), ...{ width: width } }}
          {...restProps}
        >
          <div className="i-dialog__close" onClick={closeDialog}>
            <Icon name="Close" />
          </div>
          <div className="i-dialog__header">{header}</div>
          <div className="i-dialog__body">{children}</div>
          <div className="i-dialog__footer">
            {footer ? footer : (
              <>
                <Button variant="outline" onClick={closeDialog}>取消</Button>
                <Button>确认</Button>
              </>
            )}
          </div>
        </div>
      </Transition>
    </>
  )

  return (
    ReactDOM.createPortal(DialogNode, dialogWrapper as HTMLElement)
  )
};

Dialog.displayName = 'Dialog';

export default Dialog;
