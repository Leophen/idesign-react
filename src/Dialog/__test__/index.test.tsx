import { render } from '@testing-library/react';
import Dialog from '../index';

describe('Dialog 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(
      <Dialog header="对话框标题" visible={true}>
        <span>对话框内容</span>
      </Dialog>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

