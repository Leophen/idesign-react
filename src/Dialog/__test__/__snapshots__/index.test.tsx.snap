// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog 组件测试 create 1`] = `<DocumentFragment />`;
